core/src/common.cpp

```diff
@@ -339,10 +339,16 @@ namespace irods::http
 
 	auto map_json_to_user(const nlohmann::json& _json) -> std::optional<std::string>
 	{
+		namespace logging = irods::http::log;
+
 		const auto json_res_string{to_string(_json)};
-		const static auto match_func{irods::http::globals::user_mapping_lib().get<char* (const char*)>("user_mapper_match")};
+		const static auto match_func{irods::http::globals::user_mapping_lib().get<int (const char*, char**)>("user_mapper_match")};
 		const static auto free_func{irods::http::globals::user_mapping_lib().get<void (char*)>("user_mapper_free")};
-		auto* res{match_func(json_res_string.c_str())};
+
+		char* res{};
+		if (auto rc{match_func(json_res_string.c_str(), &res)}; rc != 0) {
+			logging::error("{}: An error occured when attempting to match with error code [{}]", __func__, rc);
+		}
 
 		if (nullptr != res) {
```

Comment:
Is it possible for this to be `true` AND `rc` (above) to be non-zero?

Reply:
`rc` should only signify an internal plugin error, so we may want to do more than just log an error...

For `res` to be `nullptr` means that there is no matching user for the current information and mapping method. This is expected to happen in normal execution.

Reply:
If `rc` is non-zero, let's return a nullopt.

That will protect the iRODS server from poorly implemented plugins. _I'm thinking about the case where `rc` is non-zero and the plugin decides to return a rodsadmin user._

Reply:
> If `rc` is non-zero, let's return a nullopt.
> 
> That will protect the iRODS server from poorly implemented plugins.

Should we let it fall through to the `std::nullopt` at the bottom, or return within the if block?

> _I'm thinking about the case where `rc` is non-zero and the plugin decides to return a rodsadmin user._

It could do that if the plugin is configured to do so & the JSON data given is right (i.e. matches to something). Is there something wrong with this?

Reply:
> Should we let it fall through to the `std::nullopt` at the bottom, or return within the if block?

I think return immediately.

> It could do that if the plugin is configured to do so & the JSON data given is right (i.e. matches to something). Is there something wrong with this?

Isn't a non-zero code meant to communicate an error occurred and that the results should be ignored?

If that's not true, please explain the relationship between the error code and out pointer.

Reply:
Pardon, misread the comment.

> Isn't a non-zero code meant to communicate an error occurred and that the results should be ignored?

Yes, that is true. Was still thinking in terms of the current plugin implementation, vs defensive against a potentially problematic plugin process.

Reply:
Done.